### ISSUE TITLE
Abstract custom filter definitions

### ISSUE THREAD

**Issue by brampling** [kubeshop/botkube#657]
Currently custom [filter definitions](https://botkube.io/filters/) need to be written in Go. They then need to be called from the source code in [with_all_filters.go](https://github.com/kubeshop/botkube/blob/main/pkg/filterengine/with_all_filters.go). The BotKube container then needs to be rebuilt and redeployed.

It would be useful if custom filters could be defined in some abstracted manner (e.g. YAML).

It would also be useful if filter definitions could be kept in a ConfigMap and dynamically loaded.

This would allow:

- Non-developers to build filters
- Filters to be built faster with less boilerplate code
- Users to add filters who do not have access to rebuild/redeploy the BotKube back end

**Comment by brampling** [comment]
#412 related

**Comment by mszostok** [comment]
Initial thoughts:

IMO it would be good to extract the whole K8s source as external “plugin” first and have those filters not as a BotKube functionality but K8s source related **(if still needed).**

In case of BotKube, we can come up with some general mutation/validation webhooks approach.
```yaml
mutators:  # executed in a given order
  # run image
  - image: gcr.io/kpt-fn/set-labels:v0.1.0
    parameters:
      animal: fish
      name: herring
validators:  # executed in parallel after all mutators
  # or instead of image, call external service
  - url: example-service.default.svc.cluster.local:9111/v1/validate
```

As a result, BotKube core logic will be fairly simple and more “generic” as it can work with any sources (not only k8s).

Additional thing is that the webhooks should be connected with a given “event”. Probably we should define [CloudEvents](https://cloudevents.io/)  schema to have some common fields on all sources’ events and require from them to sent event name e.g. `<source_name>.<event_type>.<event_name>`. 

In case of k8s, the event name could be `kubernetes.resources.pod`. In that way we could add mutators/validators into the exact event (`kubernetes.resources.pod `) or for a group of them (`kubernetes.resources.*`).

It’s a draft idea, but I hope that you see the direction. Additionally, maybe it should work in both ways, not only `sources -> webhooks -> communication platform` but also `communication platform -> webhooks -> executors` - but this sounds like a feature that we don’t need right now.

**Comment by pkosiec** [comment]
@mszostok I agree, I like the external validator / mutator idea. It is also implemented in [OathKeeper from ORY](https://www.ory.sh/docs/oathkeeper/pipeline/mutator). We can also think about some built-ins if necessary.

When it comes to Docker images, I think it's a good idea (I've seen that in [keptn](https://keptn.sh/)), but we can implement it later, after the validation/mutation webhooks support, once we know if people want that 🙂

**Comment by pkosiec** [comment]
Another related issue https://github.com/kubeshop/botkube/issues/835, where @bygui86 described filtering based on messages regex. This should be possible with the new filtering functionality.

**Comment by bygui86** [comment]
@pkosiec thanks for linking my issue.

@mszostok I like to keep BotKube generic and agnostic, but the solution you proposed is pretty complex and not aligned to what BotKube currently is in my opinion. First of if you don't find a pre-built/containerised filter like `gcr.io/kpt-fn/set-labels`, then you have to write it yourself and it won't align with some important requirements mentioned by @brampling at the beginning:
- Non-developers to build filters
- Filters to be built faster with less boilerplate code
- Users to add filters who do not have access to rebuild/redeploy the BotKube back end

Secondly why should we define mutators or validators in BotKube? There are already great tools dedicated to such purpose, let's use them. However mutators and validators are neither filters on events nor checkers to come up with recommendations, so another question would be why webhooks for something "relatively simple" like filtering?

Please review my proposal in issue #835 to filter events based on matching regex on `reason` and/or `message`.
Same as "source names", the field to perform the regex match on could be flexible, adapting to any kind of JSON structure. This way BotKube can stay flexible without being k8s compatible only, even if K8s is mentioned in the description on GitHub and botkube.io...

I suggest to implement different kind of filters:
1. simple regex based matching the content of the event, defined in YAML config
2. rego syntax based, defined in YAML config

**Comment by pkosiec** [comment]
Hey @bygui86 you're right. Filtering by event messages/reasons should be a simple built-in feature without need to define/include some external filter. I'm going to reopen https://github.com/kubeshop/botkube/issues/835 and implement it in 0.16 which will come at the end of November / early December 🎉 Actually I have a draft PR prepared (https://github.com/kubeshop/botkube/pull/854), but still needs some polish. Cheers!

**Comment by bygui86** [comment]
@pkosiec thanks! Following this issue anyway as still interesting long-term :)

**Comment by pkosiec** [comment]
We're putting the `needs-triage` label as we have some basic filters in place (requested multiple times by the community).
@brampling Do we still need it? If yes, please provide us a list of requirements, and we'll prepare a draft design for it. Thanks in advance!

**Comment by brampling** [comment]
I want to keep this on hold for now and see what the feedback is about filters/actions. We still can't create custom recommendations without code modification, although event filtering is much better now. We should keep the issue open for community feedback but not bring it into the project right now.
